diff -uNr linux-6.6.1-p1/arch/sparc/mm/init_64.c linux-6.6.1-p2/arch/sparc/mm/init_64.c
--- linux-6.6.1-p1/arch/sparc/mm/init_64.c	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/arch/sparc/mm/init_64.c	2023-11-14 08:43:07.436300909 -0300
@@ -3085,11 +3085,11 @@
 static void __init kernel_lds_init(void)
 {
 	code_resource.start = compute_kern_paddr(_text);
-	code_resource.end   = compute_kern_paddr(_etext - 1);
+	code_resource.end   = compute_kern_paddr(_etext) - 1;
 	data_resource.start = compute_kern_paddr(_etext);
-	data_resource.end   = compute_kern_paddr(_edata - 1);
+	data_resource.end   = compute_kern_paddr(_edata) - 1;
 	bss_resource.start  = compute_kern_paddr(__bss_start);
-	bss_resource.end    = compute_kern_paddr(_end - 1);
+	bss_resource.end    = compute_kern_paddr(_end) - 1;
 }
 
 static int __init report_memory(void)
diff -uNr linux-6.6.1-p1/drivers/base/firmware_loader/main.c linux-6.6.1-p2/drivers/base/firmware_loader/main.c
--- linux-6.6.1-p1/drivers/base/firmware_loader/main.c	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/drivers/base/firmware_loader/main.c	2023-11-14 08:43:07.469301598 -0300
@@ -866,6 +866,11 @@
 
 	ret = _request_firmware_prepare(&fw, name, device, buf, size,
 					offset, opt_flags);
+
+#ifdef CONFIG_GENTOO_PRINT_FIRMWARE_INFO
+        printk(KERN_NOTICE "Loading firmware: %s\n", name);
+#endif
+
 	if (ret <= 0) /* error or already assigned */
 		goto out;
 
diff -uNr linux-6.6.1-p1/drivers/hwmon/Kconfig linux-6.6.1-p2/drivers/hwmon/Kconfig
--- linux-6.6.1-p1/drivers/hwmon/Kconfig	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/drivers/hwmon/Kconfig	2023-11-14 08:43:07.451301222 -0300
@@ -2137,6 +2137,7 @@
 config SENSORS_TMP513
 	tristate "Texas Instruments TMP513 and compatibles"
 	depends on I2C
+	select REGMAP_I2C
 	help
 	  If you say yes here you get support for Texas Instruments TMP512,
 	  and TMP513 temperature and power supply sensor chips.
diff -uNr linux-6.6.1-p1/fs/namei.c linux-6.6.1-p2/fs/namei.c
--- linux-6.6.1-p1/fs/namei.c	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/fs/namei.c	2023-11-14 08:43:07.431300805 -0300
@@ -1028,10 +1028,10 @@
 		path_put(&last->link);
 }
 
-static int sysctl_protected_symlinks __read_mostly;
-static int sysctl_protected_hardlinks __read_mostly;
-static int sysctl_protected_fifos __read_mostly;
-static int sysctl_protected_regular __read_mostly;
+static int sysctl_protected_symlinks __read_mostly = 1;
+static int sysctl_protected_hardlinks __read_mostly = 1;
+int sysctl_protected_fifos __read_mostly = 1;
+int sysctl_protected_regular __read_mostly = 1;
 
 #ifdef CONFIG_SYSCTL
 static struct ctl_table namei_sysctls[] = {
diff -uNr linux-6.6.1-p1/kernel/trace/bpf_trace.c linux-6.6.1-p2/kernel/trace/bpf_trace.c
--- linux-6.6.1-p1/kernel/trace/bpf_trace.c	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/kernel/trace/bpf_trace.c	2023-11-14 08:43:07.457301347 -0300
@@ -1042,7 +1042,7 @@
 };
 
 #ifdef CONFIG_X86_KERNEL_IBT
-static unsigned long get_entry_ip(unsigned long fentry_ip)
+static unsigned long __maybe_unused get_entry_ip(unsigned long fentry_ip)
 {
 	u32 instr;
 
diff -uNr linux-6.6.1-p1/mm/Kconfig linux-6.6.1-p2/mm/Kconfig
--- linux-6.6.1-p1/mm/Kconfig	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/mm/Kconfig	2023-11-14 08:43:07.441301014 -0300
@@ -1246,7 +1246,7 @@
        def_bool n
 
 config PER_VMA_LOCK
-	def_bool y
+	bool "Enable per-vma locking during page fault handling."
 	depends on ARCH_SUPPORTS_PER_VMA_LOCK && MMU && SMP
 	help
 	  Allow per-vma locking during page fault handling.
diff -uNr linux-6.6.1-p1/net/bluetooth/hci_conn.c linux-6.6.1-p2/net/bluetooth/hci_conn.c
--- linux-6.6.1-p1/net/bluetooth/hci_conn.c	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/net/bluetooth/hci_conn.c	2023-11-14 08:43:07.446301118 -0300
@@ -2342,8 +2342,13 @@
 		return 0;
 	}
 
-	if (hci_conn_ssp_enabled(conn) &&
-	    !test_bit(HCI_CONN_ENCRYPT, &conn->flags))
+	/* If Secure Simple Pairing is not enabled, then legacy connection
+	 * setup is used and no encryption or key sizes can be enforced.
+	 */
+	if (!hci_conn_ssp_enabled(conn))
+		return 1;
+
+	if (!test_bit(HCI_CONN_ENCRYPT, &conn->flags))
 		return 0;
 
 	return 1;
diff -uNr linux-6.6.1-p1/scripts/sign-file.c linux-6.6.1-p2/scripts/sign-file.c
--- linux-6.6.1-p1/scripts/sign-file.c	2023-11-08 07:56:25.000000000 -0300
+++ linux-6.6.1-p2/scripts/sign-file.c	2023-11-14 08:43:07.462301451 -0300
@@ -48,9 +48,10 @@
  * signing with anything other than SHA1 - so we're stuck with that if such is
  * the case.
  */
-#if defined(LIBRESSL_VERSION_NUMBER) || \
-	OPENSSL_VERSION_NUMBER < 0x10000000L || \
-	defined(OPENSSL_NO_CMS)
+#if defined(OPENSSL_NO_CMS) || \
+	( defined(LIBRESSL_VERSION_NUMBER) \
+	&& (LIBRESSL_VERSION_NUMBER < 0x3010000fL) ) || \
+	OPENSSL_VERSION_NUMBER < 0x10000000L
 #define USE_PKCS7
 #endif
 #ifndef USE_PKCS7
