diff -uNr linux-5.11.1/drivers/hwmon/Kconfig linux-5.11.1-p1/drivers/hwmon/Kconfig
--- linux-5.11.1/drivers/hwmon/Kconfig	2021-02-23 09:55:57.000000000 -0300
+++ linux-5.11.1-p1/drivers/hwmon/Kconfig	2021-02-26 11:49:20.305100114 -0300
@@ -1909,6 +1909,7 @@
 config SENSORS_TMP513
 	tristate "Texas Instruments TMP513 and compatibles"
 	depends on I2C
+	select REGMAP_I2C
 	help
 	  If you say yes here you get support for Texas Instruments TMP512,
 	  and TMP513 temperature and power supply sensor chips.
diff -uNr linux-5.11.1/include/uapi/linux/xattr.h linux-5.11.1-p1/include/uapi/linux/xattr.h
--- linux-5.11.1/include/uapi/linux/xattr.h	2021-02-23 09:55:57.000000000 -0300
+++ linux-5.11.1-p1/include/uapi/linux/xattr.h	2021-02-26 11:49:20.293099840 -0300
@@ -81,5 +81,9 @@
 #define XATTR_POSIX_ACL_DEFAULT  "posix_acl_default"
 #define XATTR_NAME_POSIX_ACL_DEFAULT XATTR_SYSTEM_PREFIX XATTR_POSIX_ACL_DEFAULT
 
+/* User namespace */
+#define XATTR_PAX_PREFIX XATTR_USER_PREFIX "pax."
+#define XATTR_PAX_FLAGS_SUFFIX "flags"
+#define XATTR_NAME_PAX_FLAGS XATTR_PAX_PREFIX XATTR_PAX_FLAGS_SUFFIX
 
 #endif /* _UAPI_LINUX_XATTR_H */
diff -uNr linux-5.11.1/mm/shmem.c linux-5.11.1-p1/mm/shmem.c
--- linux-5.11.1/mm/shmem.c	2021-02-23 09:55:57.000000000 -0300
+++ linux-5.11.1-p1/mm/shmem.c	2021-02-26 11:49:20.295099886 -0300
@@ -3280,6 +3280,14 @@
 	struct shmem_inode_info *info = SHMEM_I(inode);
 
 	name = xattr_full_name(handler, name);
+
+	if (!strncmp(name, XATTR_USER_PREFIX, XATTR_USER_PREFIX_LEN)) {
+		if (strcmp(name, XATTR_NAME_PAX_FLAGS))
+			return -EOPNOTSUPP;
+		if (size > 8)
+			return -EINVAL;
+	}
+
 	return simple_xattr_set(&info->xattrs, name, value, size, flags, NULL);
 }
 
@@ -3295,6 +3303,12 @@
 	.set = shmem_xattr_handler_set,
 };
 
+static const struct xattr_handler shmem_user_xattr_handler = {
+	.prefix = XATTR_USER_PREFIX,
+	.get = shmem_xattr_handler_get,
+	.set = shmem_xattr_handler_set,
+};
+
 static const struct xattr_handler *shmem_xattr_handlers[] = {
 #ifdef CONFIG_TMPFS_POSIX_ACL
 	&posix_acl_access_xattr_handler,
@@ -3302,6 +3316,7 @@
 #endif
 	&shmem_security_xattr_handler,
 	&shmem_trusted_xattr_handler,
+	&shmem_user_xattr_handler,
 	NULL
 };
 
diff -uNr linux-5.11.1/net/bluetooth/hci_conn.c linux-5.11.1-p1/net/bluetooth/hci_conn.c
--- linux-5.11.1/net/bluetooth/hci_conn.c	2021-02-23 09:55:57.000000000 -0300
+++ linux-5.11.1-p1/net/bluetooth/hci_conn.c	2021-02-26 11:49:20.300100000 -0300
@@ -1352,8 +1352,13 @@
 		return 0;
 	}
 
-	if (hci_conn_ssp_enabled(conn) &&
-	    !test_bit(HCI_CONN_ENCRYPT, &conn->flags))
+	/* If Secure Simple Pairing is not enabled, then legacy connection
+	 * setup is used and no encryption or key sizes can be enforced.
+	 */
+	if (!hci_conn_ssp_enabled(conn))
+		return 1;
+
+	if (!test_bit(HCI_CONN_ENCRYPT, &conn->flags))
 		return 0;
 
 	return 1;
diff -uNr linux-5.11.1/scripts/sign-file.c linux-5.11.1-p1/scripts/sign-file.c
--- linux-5.11.1/scripts/sign-file.c	2021-02-23 09:55:57.000000000 -0300
+++ linux-5.11.1-p1/scripts/sign-file.c	2021-02-26 11:49:20.310100229 -0300
@@ -41,9 +41,10 @@
  * signing with anything other than SHA1 - so we're stuck with that if such is
  * the case.
  */
-#if defined(LIBRESSL_VERSION_NUMBER) || \
-	OPENSSL_VERSION_NUMBER < 0x10000000L || \
-	defined(OPENSSL_NO_CMS)
+#if defined(OPENSSL_NO_CMS) || \
+	( defined(LIBRESSL_VERSION_NUMBER) \
+	&& (LIBRESSL_VERSION_NUMBER < 0x3010000fL) ) || \
+	OPENSSL_VERSION_NUMBER < 0x10000000L
 #define USE_PKCS7
 #endif
 #ifndef USE_PKCS7
