diff -uNr linux-4.14.1-acrux-r4/arch/x86/Makefile linux-4.14.1-acrux-r5/arch/x86/Makefile
--- linux-4.14.1-acrux-r4/arch/x86/Makefile	2017-11-21 16:37:53.239568566 -0200
+++ linux-4.14.1-acrux-r5/arch/x86/Makefile	2017-11-21 16:38:16.274104132 -0200
@@ -256,7 +256,7 @@
 LDFLAGS := -m elf_$(UTS_MACHINE)
 
 # Speed up the build
-KBUILD_CFLAGS += -pipe
+KBUILD_CFLAGS += -falign-functions=1 -falign-jumps=1 -falign-loops=1 -falign-labels=1 -fno-builtin -pipe
 # Workaround for a gcc prelease that unfortunately was shipped in a suse release
 KBUILD_CFLAGS += -Wno-sign-compare
 #
