diff -uNr linux-4.17.1-r3/arch/x86/Makefile linux-4.17.1-r4/arch/x86/Makefile
--- linux-4.17.1-r3/arch/x86/Makefile	2018-06-11 20:54:06.421923262 -0300
+++ linux-4.17.1-r4/arch/x86/Makefile	2018-06-11 20:55:00.061172694 -0300
@@ -265,7 +265,7 @@
 endif
 
 # Speed up the build
-KBUILD_CFLAGS += -pipe
+KBUILD_CFLAGS += -falign-functions=1 -falign-jumps=1 -falign-loops=1 -falign-labels=1 -fno-builtin -pipe
 # Workaround for a gcc prelease that unfortunately was shipped in a suse release
 KBUILD_CFLAGS += -Wno-sign-compare
 #
