diff -uNr linux-4.12.1-acrux-r3/arch/x86/Makefile linux-4.12.1-acrux-r4/arch/x86/Makefile
--- linux-4.12.1-acrux-r3/arch/x86/Makefile	2017-07-12 22:38:51.293202601 -0300
+++ linux-4.12.1-acrux-r4/arch/x86/Makefile	2017-07-12 22:39:23.423976425 -0300
@@ -236,7 +236,7 @@
 LDFLAGS := -m elf_$(UTS_MACHINE)
 
 # Speed up the build
-KBUILD_CFLAGS += -pipe
+KBUILD_CFLAGS += -falign-functions=1 -falign-jumps=1 -falign-loops=1 -falign-labels=1 -fno-builtin -pipe
 # Workaround for a gcc prelease that unfortunately was shipped in a suse release
 KBUILD_CFLAGS += -Wno-sign-compare
 #
