diff -uNr linux-4.14.14-acrux-r3/arch/x86/Makefile linux-4.14.14-acrux-r4/arch/x86/Makefile
--- linux-4.14.14-acrux-r3/arch/x86/Makefile	2018-01-17 07:12:38.524261943 -0200
+++ linux-4.14.14-acrux-r4/arch/x86/Makefile	2018-01-17 07:14:32.205024986 -0200
@@ -256,7 +256,7 @@
 LDFLAGS := -m elf_$(UTS_MACHINE)
 
 # Speed up the build
-KBUILD_CFLAGS += -pipe
+KBUILD_CFLAGS += -falign-functions=1 -falign-jumps=1 -falign-loops=1 -falign-labels=1 -fno-builtin -pipe
 # Workaround for a gcc prelease that unfortunately was shipped in a suse release
 KBUILD_CFLAGS += -Wno-sign-compare
 #
