diff -uNr linux-4.16.2-r3/arch/x86/Makefile linux-4.16.2-r4/arch/x86/Makefile
--- linux-4.16.2-r3/arch/x86/Makefile	2018-04-17 13:18:24.661827550 -0300
+++ linux-4.16.2-r4/arch/x86/Makefile	2018-04-17 13:19:15.765029319 -0300
@@ -262,7 +262,7 @@
 endif
 
 # Speed up the build
-KBUILD_CFLAGS += -pipe
+KBUILD_CFLAGS += -falign-functions=1 -falign-jumps=1 -falign-loops=1 -falign-labels=1 -fno-builtin -pipe
 # Workaround for a gcc prelease that unfortunately was shipped in a suse release
 KBUILD_CFLAGS += -Wno-sign-compare
 #
