diff -uNr linux-4.19.1-p4/drivers/pci/rom.c linux-4.19.1-p5/drivers/pci/rom.c
--- linux-4.19.1-p4/drivers/pci/rom.c	2018-11-04 11:50:54.000000000 -0200
+++ linux-4.19.1-p5/drivers/pci/rom.c	2018-11-04 18:57:44.428553119 -0200
@@ -92,8 +92,11 @@
 		void __iomem *pds;
 		/* Standard PCI ROMs start out with these bytes 55 AA */
 		if (readw(image) != 0xAA55) {
-			pci_info(pdev, "Invalid PCI ROM header signature: expecting 0xaa55, got %#06x\n",
-				 readw(image));
+			/* Some AMD PCI ROMs start out with bytes FF FF */
+			if (readw(image) != 0xFFFF) {
+				pci_info(pdev, "Invalid PCI ROM header signature: expecting 0xaa55, got %#06x\n",
+					 readw(image));
+			}
 			break;
 		}
 		/* get the PCI data structure and check its "PCIR" signature */
