diff -uNr linux-4.14.14-acrux-r4/drivers/pci/rom.c linux-4.14.14-acrux-r5/drivers/pci/rom.c
--- linux-4.14.14-acrux-r4/drivers/pci/rom.c	2018-01-17 06:45:30.000000000 -0200
+++ linux-4.14.14-acrux-r5/drivers/pci/rom.c	2018-01-17 07:16:10.829419177 -0200
@@ -92,8 +92,11 @@
 		void __iomem *pds;
 		/* Standard PCI ROMs start out with these bytes 55 AA */
 		if (readw(image) != 0xAA55) {
-			dev_err(&pdev->dev, "Invalid PCI ROM header signature: expecting 0xaa55, got %#06x\n",
-				readw(image));
+			/* Some AMD PCI ROMs start out with these bytes FF FF */
+			if (readw(image) != 0xFFFF) {
+				dev_err(&pdev->dev, "Invalid PCI ROM header signature: expecting 0xaa55, got %#06x\n",
+					readw(image));
+			}
 			break;
 		}
 		/* get the PCI data structure and check its "PCIR" signature */
