diff -uNr linux-4.17.1-r4/drivers/pci/rom.c linux-4.17.1-r5/drivers/pci/rom.c
--- linux-4.17.1-r4/drivers/pci/rom.c	2018-06-11 17:43:19.000000000 -0300
+++ linux-4.17.1-r5/drivers/pci/rom.c	2018-06-11 20:55:41.370134913 -0300
@@ -91,8 +91,11 @@
 		void __iomem *pds;
 		/* Standard PCI ROMs start out with these bytes 55 AA */
 		if (readw(image) != 0xAA55) {
-			pci_info(pdev, "Invalid PCI ROM header signature: expecting 0xaa55, got %#06x\n",
-				 readw(image));
+			/* Some AMD PCI ROMs start out with bytes FF FF */
+			if (readw(image) != 0xFFFF) {
+				pci_info(pdev, "Invalid PCI ROM header signature: expecting 0xaa55, got %#06x\n",
+					 readw(image));
+			}
 			break;
 		}
 		/* get the PCI data structure and check its "PCIR" signature */
