diff -uNr linux-5.13.1-p3/drivers/gpu/drm/amd/pm/powerplay/smumgr/ci_smumgr.c linux-5.13.1-p4/drivers/gpu/drm/amd/pm/powerplay/smumgr/ci_smumgr.c
--- linux-5.13.1-p3/drivers/gpu/drm/amd/pm/powerplay/smumgr/ci_smumgr.c	2021-07-07 09:24:58.000000000 -0300
+++ linux-5.13.1-p4/drivers/gpu/drm/amd/pm/powerplay/smumgr/ci_smumgr.c	2021-07-11 19:45:40.613708677 -0300
@@ -217,7 +217,9 @@
 
 	ret = PHM_READ_FIELD(hwmgr->device, SMC_RESP_0, SMC_RESP);
 
-	if (ret != 1)
+	if (ret == 0xFE)
+		pr_debug("message %x was not supported\n", msg);
+	else if (ret != 1)
 		pr_info("\n failed to send message %x ret is %d\n",  msg, ret);
 
 	return 0;
