diff -uNr linux-5.7.1-p3/drivers/gpu/drm/amd/powerplay/smumgr/ci_smumgr.c linux-5.7.1-p4/drivers/gpu/drm/amd/powerplay/smumgr/ci_smumgr.c
--- linux-5.7.1-p3/drivers/gpu/drm/amd/powerplay/smumgr/ci_smumgr.c	2020-06-07 06:33:02.000000000 -0300
+++ linux-5.7.1-p4/drivers/gpu/drm/amd/powerplay/smumgr/ci_smumgr.c	2020-06-08 13:15:09.099945055 -0300
@@ -216,7 +216,9 @@
 
 	ret = PHM_READ_FIELD(hwmgr->device, SMC_RESP_0, SMC_RESP);
 
-	if (ret != 1)
+	if (ret == 0xFE)
+		pr_debug("message %x was not supported\n", msg);
+	else if (ret != 1)
 		pr_info("\n failed to send message %x ret is %d\n",  msg, ret);
 
 	return 0;
